ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/helpers/OfferingToStateMapper.kt

```diff
@@ -1,31 +1,101 @@
 package com.revenuecat.purchases.ui.revenuecatui.helpers
 
+import androidx.compose.material.Colors
 import com.revenuecat.purchases.Offering
+import com.revenuecat.purchases.paywalls.PaywallData
 import com.revenuecat.purchases.ui.revenuecatui.PaywallViewMode
 import com.revenuecat.purchases.ui.revenuecatui.data.PaywallViewState
+import com.revenuecat.purchases.ui.revenuecatui.data.processed.PaywallTemplate
 import com.revenuecat.purchases.ui.revenuecatui.data.processed.TemplateConfigurationFactory
 import com.revenuecat.purchases.ui.revenuecatui.data.processed.VariableDataProvider
+import com.revenuecat.purchases.ui.revenuecatui.data.processed.VariableProcessor
+import com.revenuecat.purchases.ui.revenuecatui.errors.PaywallValidationError
+import com.revenuecat.purchases.ui.revenuecatui.extensions.createDefault
+import com.revenuecat.purchases.ui.revenuecatui.extensions.defaultTemplate
+import com.revenuecat.purchases.ui.revenuecatui.utils.Result
+
+internal fun Offering.validatedPaywall(
+    packageName: String,
+    currentColors: Colors,
+): PaywallValidationResult {
+    val paywallData = this.paywall
+        ?: return PaywallValidationResult(
+            PaywallData.createDefault(packages = availablePackages, packageName, currentColors),
+            PaywallData.defaultTemplate,
+            PaywallValidationError.MissingPaywall,
+        )
+
+    return when (val result = paywallData.validate()) {
+        is Result.Error -> PaywallValidationResult(
+            PaywallData.createDefault(packages = availablePackages, packageName, currentColors),
```

Comment:
This is the important part: instead of `availablePackages`, this should use `paywallData.packages`. So even though the paywall wasn't valid, we can still show only the packages that the developer wanted to display.